go/vt/vttablet/tabletserver/schema/historian.go

```diff
@@ -217,13 +239,25 @@ func (h *historian) readRow(row []sqltypes.Value) (*trackedSchema, int64, error)
 		tables[t.Name] = t
 	}
 	tSchema := &trackedSchema{
-		schema: tables,
-		pos:    pos,
-		ddl:    ddl,
+		schema:      tables,
+		pos:         pos,
+		ddl:         ddl,
+		timeUpdated: timeUpdated,
 	}
 	return tSchema, id, nil
 }
 
+func (h *historian) purgeOldSchemas() {
```

Comment:
I think we should do an initial pass to check if there are older schemas that need to be purged and only then create the filtered slice. Otherwise we will be recreating the same slice for a new ddl, if there is no older version to purge.

Reply:
Were you thinking something like this? 

```diff
diff --git a/go/vt/vttablet/tabletserver/schema/historian.go b/go/vt/vttablet/tabletserver/schema/historian.go
index 3099408dba..a024b0068c 100644
--- a/go/vt/vttablet/tabletserver/schema/historian.go
+++ b/go/vt/vttablet/tabletserver/schema/historian.go
@@ -248,8 +248,23 @@ func (h *historian) readRow(row []sqltypes.Value) (*trackedSchema, int64, error)
 }
 
 func (h *historian) purgeOldSchemas() {
-       filtered := make([]*trackedSchema, 0)
        maxAgeDuration := time.Duration(h.schemaMaxAgeSeconds) * time.Second
+       shouldPurge := false
+
+       // check if we have any schemas we need to purge and only create the filtered
+       // slice if necessary
+       for _, s := range h.schemas {
+               if time.Since(time.Unix(s.timeUpdated, 0)) > maxAgeDuration {
+                       shouldPurge = true
+                       break
+               }
+       }
+
+       if !shouldPurge {
+               return
+       }
+
+       filtered := make([]*trackedSchema, 0)
        for _, s := range h.schemas {
                if time.Since(time.Unix(s.timeUpdated, 0)) < maxAgeDuration {
                        filtered = append(filtered, s)
```

Reply:
Done in https://github.com/vitessio/vitess/pull/13056/commits/2a787baf81885e6082caf6b912bc056570a5a5c4